src/lib/comms/include/sol-mqtt.h

```diff
@@ -0,0 +1,393 @@
+     * broker to the client.
+     */
+    time_t keepalive;
+
+    /**
+     * NULL terminated string that should be used as client ID. If
+     * not set, clean_session must be set to true.
+     */
+    char *client_id;
+
+    /**
+     * A message that the broker should send when the client disconnects.
+     */
+    struct sol_mqtt_message *will;
+
+    /**
+     * @brief On connect callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a connect request has been processed
+     */
+    void (*connect) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On disconnect callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when the client has disconnected from the
+     * broker.
+     */
+    void (*disconnect) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On publish callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a publish request has been processed.
+     */
+    void (*publish) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On message callback
+     *
+     * @param mqtt MQTT Object
+     * @param message Message received from the broker
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a message is received from the broker.
+     * This does not include PING messages, only messages incoming due
+     * to publish request from other clients or the broker itself.
+     *
+     * The memory associated to the message object will be freed after
+     * the callback returns.
+     */
+    void (*message) (struct sol_mqtt *mqtt, const struct sol_mqtt_message *message, void *data);
+
+    /**
+     * @brief On subscribe callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a subscribe request has been processed.
+     */
+    void (*subscribe) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On unsubscribe callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a unsubscribe request has been processed.
+     */
+    void (*unsubscribe) (struct sol_mqtt *mqtt, void *data);
+};
+
+/**
+ * @brief Connect to a MQTT broker
+ *
+ * @param host The host address of the MQTT broker
+ * @param port The host port to connect to
+ * @param config Configuration and callbacks
+ * @param data User data provided to the callbacks
+ *
+ * @return New mqtt object on sucess, NULL otherwise
+ */
+struct sol_mqtt *sol_mqtt_connect(const char *host, int port, const struct sol_mqtt_config *config, void *data);
```

Comment:
const void data...

Also, I/O configs are passing the user data inside the config itself.. so better move this pointer there.